### PR TITLE
Add the planning module (updated PR)

### PR DESCRIPTION
**Pull request by danielabrozzoni**
This PR builds on top of #481, fixing all the review comments.

I didn't squash my last commits on purpose to make review easier, I can squash them before merging if preferred.

### CODE REVIEW

**Review by tcharding — APPROVED** [review]
Looks good!

Code reveiw only, I'm not super knowledgeable on this code base.

ACK 4400f85eef6b6bc6050fb0a8ccd0a057a0edaf31

**Comment by tcharding** [comment]
Looks like this is waiting for review please @sanket1729 and/or @apoelstra

File: src/plan.rs
Comment (danielabrozzoni) [reply src/plan.rs:584]: Ohh, that's surprising yes. Thanks!

**Comment by danielabrozzoni** [comment]
d097867e419f525b3717b19b83ba3a3083065e55 - rebased and fixed review comments. CI is failing due to (I think) unrelated reasons, it's broken on master as well.

**Comment by tcharding** [comment]
https://github.com/rust-bitcoin/rust-miniscript/pull/597 should fix CI

**Comment by danielabrozzoni** [comment]
Nice, rebased!

**Comment by apoelstra** [comment]
Can you run through each commit and make sure that `cargo test` passes?

**Comment by danielabrozzoni** [comment]
> Can you run through each commit and make sure that cargo test passes?

Should be cleaner now, sorry!

CI is failing, but #598 should fix it.

**Comment by danielabrozzoni** [comment]
a358076 - rebased on master

File: src/descriptor/mod.rs
Comment (apoelstra) [inline src/descriptor/mod.rs:528]: In d29c2984e60e41e3872ac23b8b59e8b25fea88a2:

In a followup PR, these are easy candidates to un-recursify (using iter.rs from #567 which was created long after this PR but merged long before).

File: src/miniscript/satisfy.rs
Comment (apoelstra) [inline src/miniscript/satisfy.rs:112]: In d29c2984e60e41e3872ac23b8b59e8b25fea88a2:

nit: I would suggest
```
NOTE: this method must ONLY allow time-based or height-based relative timelocks, except for the value
zero which is valid as both a height or a time. Allowing both could cause miniscript to consruct an invalid
witness for descriptors which mix height- and time-based timelocks. (We do not consider such descriptors
to be "sane" but they can still be processed by this library, e.g. by using `parse_insane` to parse them.)
```

And same below (but change "relative" to "absolute").

But I'm not certain about the zero comment. Can somebody more familiar with timelocks check on that?

Comment (apoelstra) [reply src/miniscript/satisfy.rs:112]: My main concern with the existing comment is that it doesn't make it clear that absolute and relative timelocks are considered separately; so you can have a height-based relative timelock and a time-based absolute timelock, for example.

Comment (apoelstra) [reply src/miniscript/satisfy.rs:112]: Also, these comments are currently on `Satisfier::check_older` and `Satisfier::check_after`. Should they also be on the equivalent methods in `AssetProvider`?

File: src/plan.rs
Comment (apoelstra) [inline src/plan.rs:248]: In d29c2984e60e41e3872ac23b8b59e8b25fea88a2:

Maybe this was discussed in the other thread, but should we make this generic over `Pk` rather than using a `DefiniteDescriptorKey`?

**Comment by apoelstra** [comment]
a358076cb88fe51952abf79d5072eb2f30aeadd3 looks amazing! I appreciate the detailed type-guided API with structs like `TaprootCanSign`.

I haven't tried to use this yet so there may be pain points but reading the code, this looks great. My comments are only nits.

**Review by apoelstra — APPROVED** [review]
ACK a358076cb88fe51952abf79d5072eb2f30aeadd3

File: src/descriptor/mod.rs
Comment (tcharding) [reply src/descriptor/mod.rs:528]: For more context, what he means is what I'm doing in https://github.com/rust-bitcoin/rust-miniscript/pull/594. The `iter` module implements a pretty hard core data structure, if you don't have the time or inclination to work it out right now just holla at me and I'll hack it up for you.

**Review by sanket1729 — APPROVED** [review]
ACK a358076cb88fe51952abf79d5072eb2f30aeadd3

I have quite a bit with this PR in Summer of bitcoin project with @Harshil-Jani who wrote some example programs using this API in #559. Thanks a lot for sticking with this PR for over a year :)

**Comment by tcharding** [comment]
Party! This one was open for a long time huh.

**Pull request merged**
